binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java

```diff
@@ -49,118 +48,142 @@ abstract class SerdeResolverUtils {
+			return true;
+		}
+		if (Serdes.Float().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		else if (Serdes.ByteArray().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		else if (Serdes.String().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
+		}
+		else if (Serdes.UUID().getClass().isAssignableFrom(serde.getClass())) {
+			return true;
 		}
 		return false;
 	}
 
 	/**
-	 * Find all {@link Serde} beans that are assignable from {@code targetType}.
+	 * Find all {@link Serde} beans that can be used for {@code targetType}.
 	 *
 	 * @param context the application context
 	 * @param targetType the target type the serdes are being matched for
-	 * @return list of matching serdes order by most specific match, or an empty list if no matches found
+	 * @return list of matching serdes ordered by most specific match, or an empty list if no matches found
 	 */
 	static List<Serde<?>> findMatchingSerdes(ConfigurableApplicationContext context, ResolvableType targetType) {
 		// We don't attempt to find a matching Serde for type '?'
 		if (targetType.getRawClass() == null) {
 			return Collections.emptyList();
 		}
 		List<SerdeWithSpecificityScore> matchingSerdes = new ArrayList<>();
-
-		context.getBeansOfType(Serde.class).forEach((beanName, serdeBean) -> {
-			final Class<?> beanConfigClass = ClassUtils.resolveClassName(((AnnotatedBeanDefinition)
-					context.getBeanFactory().getBeanDefinition(beanName))
-					.getMetadata().getClassName(),
-				ClassUtils.getDefaultClassLoader());
+		ResolvableType serdeType = ResolvableType.forClassWithGenerics(Serde.class, targetType);
+		String[] serdeBeanNames = BeanFactoryUtils.beanNamesForTypeIncludingAncestors(context.getBeanFactory(),
+			serdeType, false, false);
+		Arrays.stream(serdeBeanNames).forEach((beanName) -> {
 			try {
-				Method[] methods = beanConfigClass.getMethods();
-				Optional<Method> serdeBeanMethod = Arrays.stream(methods).filter(m -> m.getName().equals(beanName)).findFirst();
-				serdeBeanMethod.ifPresent((method) -> {
-					ResolvableType serdeBeanMethodReturnType = ResolvableType.forMethodReturnType(method, beanConfigClass);
-					ResolvableType serdeBeanGeneric = serdeBeanMethodReturnType.getGeneric(0);
-					// We don't attempt to use a Serde<?> as a match for anything currently
-					if (serdeBeanGeneric.getRawClass() != null && serdeBeanGeneric.isAssignableFrom(targetType)) {
-						matchingSerdes.add(new SerdeWithSpecificityScore(calculateScore(targetType, serdeBeanGeneric), serdeBean));
-					}
-				});
-			}
-			catch (Exception e) {
-				if (LOG.isTraceEnabled()) {
-					LOG.trace("Failed to introspect Serde bean method '" + serdeBean + "'", e);
+				BeanDefinition beanDefinition = context.getBeanFactory().getMergedBeanDefinition(beanName);
+				ResolvableType serdeBeanGeneric = beanDefinition.getResolvableType().getGeneric(0);
+				Serde<?> serdeBean = context.getBean(beanName, Serde.class);
```

Comment:
Yes, this thought crossed my mind as well as sounded something like this....

"We are not even using this ordered list of matching beans. Now we have this complicated set of tests to look at / maintain. We should only add things when we actually need them".   

1. However, we do have to go through them all in order to find the "closest/best match".
2. I wanted to test the sort/ordering from above so I ended up w/ the test permutations.

Since at this point the code was already there I chose to leave it in rather than pare it back. To limit any public use of the N-based API we could instead create a ` public Serde<?> findClosestMatchingSerde` API which delegates to the existing one and move the existing one to package-protected. 

I think this would be a good thing. Wdyt?

Reply:
Not sure how the suggested refactoring is relevant to my concern about beans loading overhead.

According the rest of your logic, you do propagate `Serde` beans in the `SerdeWithSpecificityScore`, but you do not use them in the score calculation and then you don't use the in compare function.
Not sure how dropping bean reference at this point in favor if just its name may break all your tests...

Reply:
Ahh ok. I misunderstood the concern.

Yes, since we do not actually use the bean for anything other than returning to the caller, we could instead just return the beanName(s) and let the caller load the bean(s), OR if we pared the API down to return only a single "top match" we could just load that single serde bean.

One thing to remember is that this operation is during the initial configuration so not loading the handful of serde beans probably configured will not be a big performance gain.

> Not sure how dropping bean reference at this point in favor if just its name may break all your tests..

Let's not worry about that. I will update the tests based on the direction we choose. (we tell them what to do ;P )

Reply:
Well, if it's used only on configuration phase, then OK: let's your logic a is.
Plus there is no too much to worry about since the API in the end is not `public`.